### PR TITLE
User specified FPGA kernel boundaries 

### PR DESCRIPTION
**Pull request by TizianoDeMatteis**
With this pull request, we want to allow the FPGA programmer to specify whether a given FPGA state must be code generated as an FPGA kernel or not.

To do this, we use the `location` property of the SDFG state. This has a key `is_FPGA_kernel`, that, by default is set to true. When is true, the state is inferred to be an FPGA Kernel.

Playing with this property, we can allow the generation of a separate FPGA kernel for each NestedSDFG

### CODE REVIEW

File: tests/fpga/nested_sdfg_as_kernel.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+        schedule=dace.dtypes.ScheduleType.FPGA_Device)
+
+    vecAdd_tasklet = fpga_state.add_tasklet('vecAdd_task', ['x_con', 'y_con'],
+                                            ['z_con'], 'z_con = x_con + y_con')
+
+    fpga_state.add_memlet_path(x,
+                               vecMap_entry,
+                               vecAdd_tasklet,
+                               dst_conn='x_con',
+                               memlet=dace.Memlet.simple(x.data, "i"))
```
Comment (definelicht) [inline tests/fpga/nested_sdfg_as_kernel.py:61]: `Memlet.simple` is deprecated, use `dace.Memlet(f"{x.data}[i]")`

**Review by definelicht — CHANGES_REQUESTED** [review]

File: dace/sdfg/state.py [hunk 2]
```diff
@@ -731,6 +731,8 @@ def __init__(self, label=None, sdfg=None, debuginfo=None, location=None):
         self.nosync = False
         self.location = location if location is not None else {}
         self._default_lineinfo = None
+        # By default, FPGA states are inferred as FPGA kernels
+        self.location["is_FPGA_kernel"] = True
```
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply dace/sdfg/state.py:735]: Do you mean, not setting that default value and, whenever needed, check the value of that property if it exists?

File: tests/fpga/nested_sdfg_as_kernel.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+
+import dace
+import numpy as np
+import argparse
+import subprocess
+
+from dace.memlet import Memlet
+
+
+def make_vecAdd_sdfg(dtype=dace.float32):
```
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply tests/fpga/nested_sdfg_as_kernel.py:19]: are they listed somewhere? :)

File: dace/sdfg/state.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (definelicht) [reply dace/sdfg/state.py:735]: Yes, location is a bit of a wild west in that sense

File: tests/fpga/nested_sdfg_as_kernel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (definelicht) [reply tests/fpga/nested_sdfg_as_kernel.py:19]: I believe we are following google conventions with exceptions

[the remark this answers is not in the document]
Comment (definelicht) [reply tests/fpga/nested_sdfg_as_kernel.py:19]: Also, mixing_underscores_and_camelCase is extra naughty 🤠

File: tests/fpga/multiple_kernels_stream.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                   transient=True)
+    sdfg.add_array("device_c", shape=[1], dtype=dtype, storage=dace.dtypes.StorageType.FPGA_Global,
+                   transient=True)
+
+    in_device_a = copy_in_state.add_write("device_a")
+    in_device_c = copy_in_state.add_write("device_c")
+
+    copy_in_state.add_memlet_path(
+        in_host_a, in_device_a,
+        memlet=Memlet.simple(in_host_a, "0")
```
Comment (definelicht) [inline tests/fpga/multiple_kernels_stream.py:38]: `Memlet.simple` is deprecated :-) use `Memlet(f"{in_host_a}[0]")`

File: tests/fpga/multiple_kernels_stream.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    fpga_state_0.add_memlet_path(
+        a_in, state_0_tasklet,
+        dst_conn='inCon',
+        memlet=dace.Memlet.simple(a_in.data, '0')
+    )
+
+    fpga_state_0.add_memlet_path(
+        state_0_tasklet, b_stream_out,
+        src_conn='outCon',
+        memlet=dace.Memlet.simple(b_stream_out, '0', num_accesses=-1)
```
Comment (definelicht) [inline tests/fpga/multiple_kernels_stream.py:86]: `num_accesses` is deprecated, use `dynamic=True`

File: tests/fpga/multiple_kernels_stream.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                  dace.sdfg.sdfg.InterstateEdge())
+    sdfg.add_edge(fpga_state_0, fpga_state_1,
+                  dace.sdfg.sdfg.InterstateEdge())
+    sdfg.add_edge(fpga_state_1, copy_out_state,
+                  dace.sdfg.sdfg.InterstateEdge())
+
+    #########
+    # Validate
+    sdfg.fill_scope_connectors()
+    sdfg.save('/tmp/out.sdfg')
```
Comment (definelicht) [inline tests/fpga/multiple_kernels_stream.py:129]: Remove I think?

**Review by definelicht — CHANGES_REQUESTED** [review]

**Review by definelicht — APPROVED** [review]

**Pull request merged**